File: web/src/session.ts
import type { MarkdownEditor } from './editor.js';
import type { SelectionRange } from './core.js';

export interface SessionDocument {
  id: string;
  markdown: string;
  selection: SelectionRange | null;
  touchedAt: number;
}

/** A bounded, editor-agnostic set of open documents with deterministic save/switch. */
export class MarkdownSession {
  readonly editor: MarkdownEditor;
  readonly maxDocuments: number;
  private documents = new Map<string, SessionDocument>();
  private activeId: string | null = null;

  constructor(editor: MarkdownEditor, options: { maxDocuments?: number } = {}) {
    this.editor = editor;
    this.maxDocuments = Math.max(1, options.maxDocuments ?? 16);
  }

  get activeDocumentId(): string | null { return this.activeId; }
  get openDocumentIds(): string[] { return [...this.documents.keys()]; }

  open(id: string, markdown: string): void {
    if (!id.trim()) throw new Error('A session document id must not be empty');
    this.saveActive();
    const existing = this.documents.get(id);
    const document = existing ?? { id, markdown, selection: null, touchedAt: 0 };
    document.markdown = existing?.markdown ?? markdown;
    document.touchedAt = performance.now();
    this.documents.delete(id);
    this.documents.set(id, document);
    this.activeId = id;
    this.editor.setMarkdown(document.markdown);
    if (document.selection) this.editor.setSelectionRange(document.selection);
    this.evictInactive();
  }

  switchTo(id: string): boolean {
    const document = this.documents.get(id);
    if (!document || id === this.activeId) return !!document;
    this.open(id, document.markdown);
    return true;
  }

  close(id: string): boolean {
    const existed = this.documents.delete(id);
    if (this.activeId === id) this.activeId = null;
    return existed;
  }

  snapshot(id: string): SessionDocument | null {
    if (id === this.activeId) this.saveActive();
    const value = this.documents.get(id);
    return value ? { ...value, selection: value.selection ? { ...value.selection } : null } : null;
  }

  saveActive(): void {
    if (!this.activeId) return;
    const document = this.documents.get(this.activeId);
    if (!document) return;
    document.markdown = this.editor.markdown;
    document.selection = globalThis.document.activeElement === this.editor.root
      ? this.editor.selectionRange()
      : document.selection;
    document.touchedAt = performance.now();
  }

  destroy(): void {
    this.saveActive();
    this.documents.clear();
    this.activeId = null;
  }

  private evictInactive(): void {
    while (this.documents.size > this.maxDocuments) {
      const candidate = this.documents.keys().next().value;
      if (candidate === undefined) break;
      if (candidate === this.activeId) {
        const current = this.documents.get(candidate)!;
        this.documents.delete(candidate);
        this.documents.set(candidate, current);
        continue;
      }
      this.documents.delete(candidate);
    }
  }
}

File: apple/Sources/MDEditorUI/Session.swift
import Foundation

public struct MarkdownSessionDocument: Sendable, Equatable {
    public let id: String
    public var markdown: String
    public var selection: NSRange?
    public var touchedAt: TimeInterval
}

/// Bounded open-document state that saves source and selection before every switch.
public final class MarkdownSession {
    public let editor: MarkdownTextView
    public let maxDocuments: Int
    public let maxWarmDocuments: Int
    public let maxWarmBytes: Int
    private var documents: [String: MarkdownSessionDocument] = [:]
    private var warm: [String: NSAttributedString] = [:]
    private var warmCosts: [String: Int] = [:]
    private var warmOrder: [String] = []
    private var order: [String] = []
    public private(set) var activeDocumentID: String?
    internal private(set) var projectionCaptureCount = 0

    public init(
        editor: MarkdownTextView,
        maxDocuments: Int = 16,
        maxWarmDocuments: Int = 4,
        maxWarmBytes: Int = 16 * 1024 * 1024
    ) {
        self.editor = editor
        self.maxDocuments = max(1, maxDocuments)
        self.maxWarmDocuments = max(0, min(self.maxDocuments, maxWarmDocuments))
        self.maxWarmBytes = max(0, maxWarmBytes)
    }

    public var openDocumentIDs: [String] { order }
    public var warmDocumentIDs: [String] { warmOrder }
    public var warmProjectionBytes: Int { warmCosts.values.reduce(0, +) }

    public func open(id: String, markdown: String) throws {
        guard !id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw MarkdownSessionError.emptyID
        }
        saveActive()
        var document = documents[id] ?? MarkdownSessionDocument(
            id: id, markdown: markdown, selection: nil, touchedAt: 0
        )
        document.touchedAt = Date.timeIntervalSinceReferenceDate
        documents[id] = document
        order.removeAll { $0 == id }
        order.append(id)
        activeDocumentID = id
        if let projection = warm[id],
           editor.restoreProjection(markdown: document.markdown, projection: projection) {
            warmOrder.removeAll { $0 == id }
            warmOrder.append(id)
        } else {
            editor.setMarkdown(document.markdown)
        }
        if let selection = document.selection, selection.upperBound <= editor.markdown.utf16.count {
            editor.selectedRange = selection
        }
        evictInactive()
    }

    @discardableResult
    public func switchTo(id: String) throws -> Bool {
        guard let document = documents[id] else { return false }
        if activeDocumentID != id { try open(id: id, markdown: document.markdown) }
        return true
    }

    @discardableResult
    public func close(id: String) -> Bool {
        let existed = documents.removeValue(forKey: id) != nil
        warm.removeValue(forKey: id)
        warmCosts.removeValue(forKey: id)
        warmOrder.removeAll { $0 == id }
        order.removeAll { $0 == id }
        if activeDocumentID == id { activeDocumentID = nil }
        return existed
    }

    public func snapshot(id: String) -> MarkdownSessionDocument? {
        if activeDocumentID == id { saveActive() }
        return documents[id]
    }

    public func saveActive() {
        guard let id = activeDocumentID, var document = documents[id] else { return }
        document.markdown = editor.markdown
        document.selection = editor.selectedRange
        document.touchedAt = Date.timeIntervalSinceReferenceDate
        documents[id] = document
        // A restored warm projection is already an immutable snapshot of this exact
        // source. Re-copying its entire attributed string on every switch creates
        // document-sized transient allocations even when the user changed nothing.
        // Preserve it and only capture again after the source actually diverges.
        if maxWarmDocuments > 0, warm[id]?.string == document.markdown {
            warmOrder.removeAll { $0 == id }
            warmOrder.append(id)
        } else if maxWarmDocuments > 0, let projection = editor.captureProjection() {
            projectionCaptureCount += 1
            warm[id] = projection
            // UTF-16 storage is the stable lower bound. Attribute dictionaries may
            // share values, so avoid pretending an imprecise heap estimate is exact.
            warmCosts[id] = projection.length * MemoryLayout<unichar>.stride
            warmOrder.removeAll { $0 == id }
            warmOrder.append(id)
            while warmOrder.count > maxWarmDocuments || warmProjectionBytes > maxWarmBytes {
                let victim = warmOrder.removeFirst()
                warm.removeValue(forKey: victim)
                warmCosts.removeValue(forKey: victim)
            }
        }
    }

    private func evictInactive() {
        while order.count > maxDocuments,
              let candidate = order.first(where: { $0 != activeDocumentID }) {
            documents.removeValue(forKey: candidate)
            warm.removeValue(forKey: candidate)
            warmCosts.removeValue(forKey: candidate)
            warmOrder.removeAll { $0 == candidate }
            order.removeAll { $0 == candidate }
        }
    }
}

public enum MarkdownSessionError: Error { case emptyID }
